### PR TITLE
one-hot encoder

### PR DESCRIPTION
**Pull request by sugarspectre**

### CODE REVIEW

File: src/linalg/utils.ts [hunk 1]
```diff
@@ -18,6 +18,25 @@ export const checkDimension = (matrix: Tensor, dim: number): boolean => {
...
+ * @param shape target shape
+ * @returns
+ */
+export const checkShape = (matrix: Tensor, shape: number[]): boolean => {
+  const mShape = matrix.shape;
+  if (mShape.length != shape.length) {
+    return false;
+  } else {
+    for (let i = 0; i < mShape.length; i++) {
+      if (shape[i] == -1) continue;
```
Comment (FeelyChau) [inline src/linalg/utils.ts:33]:
```suggestion
      if (shape[i] === -1) continue;
```

File: src/linalg/utils.ts [hunk 2]
```diff
@@ -18,6 +18,25 @@ export const checkDimension = (matrix: Tensor, dim: number): boolean => {
...
+      if (shape[i] != mShape[i]) return false;
```
Comment (FeelyChau) [inline src/linalg/utils.ts:34]:
```suggestion
      if (shape[i] !== mShape[i]) return false;
```

File: src/linalg/utils.ts [hunk 3]
```diff
@@ -18,6 +18,25 @@ export const checkDimension = (matrix: Tensor, dim: number): boolean => {
...
+ * Check if matrix of target shape
+ * @param matrix matrix tensor
+ * @param shape target shape
+ * @returns
+ */
+export const checkShape = (matrix: Tensor, shape: number[]): boolean => {
+  const mShape = matrix.shape;
+  if (mShape.length != shape.length) {
+    return false;
+  } else {
```
Comment (FeelyChau) [inline src/linalg/utils.ts:31]: `else` is not necessary because it returns in `if` scope.

File: src/preprocess/encoder.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  public constructor (params: OneHotEncoderParams = { drop: 'none' }) {
+    const { drop } = params;
+    this.drop = drop;
+  }
+  /**
+   * Init one-hot encoder
+   * @param x data input used to init encoder
+   * @param categories user input categories
+   */
+  public async init(x: Tensor | number[] | string[]):Promise<void> {
```
Comment (yorkie) [inline src/preprocess/encoder.ts:49]:
```suggestion
  public async init(x: Tensor | number[] | string[]): Promise<void> {
```

Comment (yorkie) [reply src/preprocess/encoder.ts:49]: Why not using TensorLike?

File: src/preprocess/encoder.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+   *    However, dropping one category may introduce a bias term in
+   *    downstream models. For instance in linear regression model,
+   *    the effect of the dropped feature might appear in intercept term.
+   *
+   *    - 'none': default, return all features
+   *    - 'first': drop the first categories in each feature
+   *    - 'binary-only': drop the first category in each feature with two
+   *       categories.
+   */
+  public constructor (params: OneHotEncoderParams = { drop: 'none' }) {
```
Comment (FeelyChau) [inline src/preprocess/encoder.ts:40]: If we define an enum for `drop`, we will get the code prompt and type check.

File: test/node/preprocess/encoder.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    assert.deepEqual(bxOneHot.dataSync(), bxEncodeDrop.dataSync());
+  });
+  it('decode binary only', async () => {
+    const encoder = new OneHotEncoder({drop: 'binary-only'});
+    await encoder.init(bx);
+    const bxCate = await encoder.decode(bxEncodeDrop);
+    // @ts-ignore
+    assert.deepEqual(bxCate.dataSync(), bx);
+  });
+});
```
Comment (FeelyChau) [inline test/node/preprocess/encoder.ts:62]: EOF.

File: src/preprocess/encoder.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    }
+  }
+  /**
+   * Decode one-hot array to original category array
+   * @param x one-hot format data need to transform
+   * @returns transformed category data
+   */
+  public async decode (x: Tensor): Promise<Tensor> {
+    const nCate = this.categories.shape[0];
+    const codeSize = this.drop == 'first' ? nCate - 1 : this.drop == 'binary-only' && nCate == 2 ? 1 : nCate;
```
Comment (FeelyChau) [inline src/preprocess/encoder.ts:92]:
```suggestion
    const codeSize = this.drop === 'first' ? nCate - 1 : this.drop === 'binary-only' && nCate === 2 ? 1 : nCate;
```

File: src/preprocess/encoder.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+   */
+  public async encode (x: Tensor | number[] | string[]): Promise<Tensor> {
+    if (!this.categories) {
+      throw Error('Please init encoder using init()');
+    } else {
+      const xTensor = checkArray(x, 'any', 1);
+      const xData = await xTensor.dataSync();
+      const nCate = this.categories.shape[0];
+      const xInd = xData.map((d: number|string) => this.cateMap[d]);
+      if (this.drop == 'binary-only' && nCate == 2) {
```
Comment (FeelyChau) [inline src/preprocess/encoder.ts:74]:
```suggestion
      if (this.drop === 'binary-only' && nCate === 2) {
```

File: src/preprocess/encoder.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        return tensor(xInd);
+      } else if (this.drop == 'first') {
```
Comment (FeelyChau) [inline src/preprocess/encoder.ts:76]:
```suggestion
      } else if (this.drop === 'first') {
```

**Review by FeelyChau — CHANGES_REQUESTED** [review]

File: src/preprocess/encoder.ts [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    this.cateMap = cateMap;
+  }
+
+  /**
+   * Encode a given feature into one-hot format
+   * @param x feature array need to encode
+   * @returns transformed one-hot feature
+   */
+  public async encode (x: Tensor | number[] | string[]): Promise<Tensor> {
```
Comment (yorkie) [inline src/preprocess/encoder.ts:66]:
```suggestion
  public async encode(x: Tensor | number[] | string[]): Promise<Tensor> {
```

File: src/preprocess/encoder.ts [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    if (!this.categories) {
+      throw Error('Please init encoder using init()');
```
Comment (yorkie) [inline src/preprocess/encoder.ts:68]:
```suggestion
      throw TypeError('Please init encoder using init()');
```

File: src/preprocess/encoder.ts (hunk 8 above)
Comment (yorkie) [inline src/preprocess/encoder.ts:74]:
```suggestion
      if (this.drop === 'binary-only' && nCate === 2) {
```
Use strict equals.

File: src/preprocess/encoder.ts (hunk 9 above)
Comment (yorkie) [inline src/preprocess/encoder.ts:76]: Ditto

File: src/preprocess/encoder.ts [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    } else {
+      const xTensor = checkArray(x, 'any', 1);
+      const xData = await xTensor.dataSync();
+      const nCate = this.categories.shape[0];
+      const xInd = xData.map((d: number|string) => this.cateMap[d]);
+      if (this.drop == 'binary-only' && nCate == 2) {
+        return tensor(xInd);
+      } else if (this.drop == 'first') {
+        const xOneHot = oneHot(cast(sub(tensor(xInd), 1), 'int32'), nCate - 1);
```
Comment (yorkie) [inline src/preprocess/encoder.ts:77]:
```suggestion
        return oneHot(cast(sub(tensor(xInd), 1), 'int32'), nCate - 1);
```

File: src/preprocess/encoder.ts [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        return xOneHot;
+      } else {
+        const xOneHot = oneHot(cast(tensor(xInd), 'int32'), nCate);
```
Comment (yorkie) [inline src/preprocess/encoder.ts:80]: Ditto.

File: src/preprocess/encoder.ts (hunk 7 above)
Comment (yorkie) [inline src/preprocess/encoder.ts:92]:
```suggestion
    const codeSize = this.drop === 'first' ? nCate - 1 : this.drop === 'binary-only' && nCate === 2 ? 1 : nCate;
```

File: src/preprocess/encoder.ts [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        return xOneHot;
+      }
+    }
+  }
+  /**
+   * Decode one-hot array to original category array
+   * @param x one-hot format data need to transform
+   * @returns transformed category data
+   */
+  public async decode (x: Tensor): Promise<Tensor> {
```
Comment (yorkie) [inline src/preprocess/encoder.ts:90]:
```suggestion
  public async decode(x: Tensor): Promise<Tensor> {
```

File: src/preprocess/encoder.ts [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    const nCate = this.categories.shape[0];
+    const codeSize = this.drop == 'first' ? nCate - 1 : this.drop == 'binary-only' && nCate == 2 ? 1 : nCate;
+    const shapeCorrect = codeSize > 1 ? checkShape(x, [ -1, codeSize ]) : checkShape(x, [ -1 ]);
+    if (!shapeCorrect) {
+      throw new TypeError('Input shape does not match');
+    }
+    const cateInd = (this.drop == 'binary-only' && nCate == 2) ? await x.data() : await argMax(x, 1).data();
+    const cateTensors: Tensor[] = [];
+    if (this.drop == 'binary-only' && nCate == 2) {
```
Comment (yorkie) [inline src/preprocess/encoder.ts:99]: Ditto

File: src/preprocess/encoder.ts [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+          throw Error('Index out of range');
+        }
+        return cateTensors.push(slice(this.categories, ind, 1));
+      });
+    } else if (this.drop == 'first') {
+      cateInd.forEach((ind: number, i: number) => {
+        if (Number(slice(x, [ i, ind ], [ 1, 1 ]).dataSync()) == 0) {
+          return cateTensors.push(slice(this.categories, 0, 1));
+        }
+        return cateTensors.push(slice(this.categories, ind + 1, 1));
```
Comment (yorkie) [inline src/preprocess/encoder.ts:111]: forEach's iterator doesn't expect a returned value.

File: src/preprocess/encoder.ts [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+      });
+    } else {
+      cateInd.forEach((ind: number) => {
+        return cateTensors.push(slice(this.categories, ind, 1));
+      });
+    }
+    const cateVal = reshape(stack(cateTensors), [ -1 ]);
```
Comment (yorkie) [inline src/preprocess/encoder.ts:118]:
```suggestion
    return reshape(stack(cateTensors), [ -1 ]);
```

File: test/node/preprocess/encoder.ts (hunk 6 above)
Comment (yorkie) [inline test/node/preprocess/encoder.ts:62]: EOF

File: test/node/preprocess/encoder.ts [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+    const encoder = new OneHotEncoder({drop: 'binary-only'});
+    await encoder.init(bx);
+    const bxOneHot = await encoder.encode(bx);
+    assert.deepEqual(bxOneHot.dataSync(), bxEncodeDrop.dataSync());
+  });
+  it('decode binary only', async () => {
+    const encoder = new OneHotEncoder({drop: 'binary-only'});
+    await encoder.init(bx);
+    const bxCate = await encoder.decode(bxEncodeDrop);
+    // @ts-ignore
```
Comment (yorkie) [inline test/node/preprocess/encoder.ts:59]: Why ignore the following line?

Comment (sugarspectre) [reply test/node/preprocess/encoder.ts:59]: use dataSync() function on a string tensor will cause type error, the return types of dataSync() are  limited to numeric arrays

Comment (yorkie) [reply test/node/preprocess/encoder.ts:59]: Use the keyword `as` to declare the mentioned case instead of `ts-ignore` which is lack of details.

**Pull request merged**